packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +104,224 @@ if (process.env.NODE_ENV === 'development') {
+      return;
+    }
+
+    try {
+      isGuessing.current = true;
+
+      // We need to wait for the ref to be attached so that we can measure available and required sizes.
+      if (containerRef.current == null) {
+        return;
+      }
+
+      const dimension = orientation === 'horizontal' ? 'Width' : 'Height';
+      const availableSize = containerRef.current[`offset${dimension}`];
+      const requiredSize = containerRef.current[`scroll${dimension}`];
+
+      const isOverflowing = availableSize < requiredSize;
+
+      if (
+        // We have already found the correct visibleCount
+        (visibleCount === itemsCount && !isOverflowing) ||
+        // if the new average of visibleCountGuessRange will never change the visibleCount anymore (infinite loop)
+        (visibleCountGuessRange[1] - visibleCountGuessRange[0] === 1 &&
+          visibleCount === visibleCountGuessRange[0])
+      ) {
+        setVisibleCountGuessRange(null);
+        return;
+      }
+
+      // Before the main logic, the max guess MUST be >= the correct visibleCount for the algorithm to work.
+      // If not:
+      // - double the max guess and visibleCount: since we need to overflow.
+      // - set min guess to current visibleCount: since underflow means correct visibleCount >= current visibleCount.
+      if (visibleCountGuessRange[1] === visibleCount && !isOverflowing) {
+        const doubleOfMaxGuess = visibleCountGuessRange[1] * 2;
+
+        setVisibleCountGuessRange([visibleCount, doubleOfMaxGuess]);
+        setVisibleCount(doubleOfMaxGuess);
+
+        return;
+      }
+
+      let newVisibleCountGuessRange = visibleCountGuessRange;
+
+      if (isOverflowing) {
+        // overflowing = we guessed too high. So, new max guess = half the current guess
+        newVisibleCountGuessRange = [visibleCountGuessRange[0], visibleCount];
+      } else {
+        // not overflowing = maybe we guessed too low? So, new min guess = half of current guess
+        newVisibleCountGuessRange = [visibleCount, visibleCountGuessRange[1]];
+      }
+
+      // Next guess is always the middle of the new guess range
+      const newVisibleCount = Math.floor(
+        (newVisibleCountGuessRange[0] + newVisibleCountGuessRange[1]) / 2,
+      );
+
+      // If newVisibleCount is 0 (impossible state as min is 1), end guessing
+      if (newVisibleCount === 0) {
+        setVisibleCountGuessRange(null);
+        return;
+      }
+
+      setVisibleCountGuessRange(newVisibleCountGuessRange);
+      setVisibleCount(newVisibleCount);
+    } finally {
+      isGuessing.current = false;
+    }
+  }, [
+    containerRef,
+    isStabilized,
+    itemsCount,
+    orientation,
+    setVisibleCount,
+    visibleCount,
+    visibleCountGuessRange,
+  ]);
+
+  const previousVisibleCount = useLatestRef(visibleCount);
+  const previousVisibleCountGuessRange = useLatestRef(visibleCountGuessRange);
+  const previousContainer = useLatestRef(containerRef.current);
```

Comment:
I'm not sure why we're using `useLatestRef` here, if we're manually changing these things in the `useLayoutEffect` anyway.

I also think these variables might not even be necessary. The `useLayoutEffect` dependency array could automatically take care of these things. To make that work, `guessVisibleCount` is what should be passed through `useLatestRef` to ensure that it doesn't re-trigger the Effect every time it changes.

And the `containerRef` shouldn't ever change. If it does, we should probably reset everything.

Reply:
> > I recently noticed that this logic does not work properly for `Select`.
> 
> I found that this was because we weren't handling the `itemsCount` change case. Now, that case is handled where an `itemsCount` change resets the guess related variables. ([6a85122](https://github.com/iTwin/iTwinUI/commit/6a85122861515002b1d8c845b92fce4ad1f47121)).

I think we can handle this using `key` as well, and not have to maintain the `previousItemsCount` state.

<details><summary>Patch</summary>

```tsx
diff --git a/packages/itwinui-react/src/utils/components/OverflowContainer.tsx b/packages/itwinui-react/src/utils/components/OverflowContainer.tsx
index 03e28d68c..36067f912 100644
--- a/packages/itwinui-react/src/utils/components/OverflowContainer.tsx
+++ b/packages/itwinui-react/src/utils/components/OverflowContainer.tsx
@@ -53,7 +53,7 @@ type OverflowContainerProps = {
  *   );
  * };
  */
-const OverflowContainerMain = React.forwardRef((props, ref) => {
+const OverflowContainerMain = React.forwardRef((props, forwardedRef) => {
   const { itemsCount, children, overflowOrientation, ...rest } = props;
 
   const [containerRef, visibleCount] = useOverflow(
@@ -68,7 +68,7 @@ const OverflowContainerMain = React.forwardRef((props, ref) => {
 
   return (
     <OverflowContainerContext.Provider value={overflowContainerContextValue}>
-      <Box ref={useMergedRefs(ref, containerRef)} {...rest}>
+      <Box ref={useMergedRefs(forwardedRef, containerRef)} {...rest}>
         {children}
       </Box>
     </OverflowContainerContext.Provider>
@@ -103,18 +103,23 @@ const OverflowContainerOverflowNode = (
 // ----------------------------------------------------------------------------
 
 const OverflowContainerComponent = React.forwardRef((props, forwardedRef) => {
-  const { overflowOrientation = 'horizontal', ...rest } = props;
+  const { itemsCount, overflowOrientation = 'horizontal', ...rest } = props;
 
   const [size, setSize] = React.useState<DOMRectReadOnly | null>(null);
   const [resizeRef] = useResizeObserver(setSize);
 
   const ref = useMergedRefs(resizeRef, forwardedRef);
 
+  const key =
+    `${itemsCount}` +
+    `${overflowOrientation === 'vertical' ? size?.height : size?.width}`;
+
   return (
     <OverflowContainerMain
       {...rest}
-      key={overflowOrientation === 'vertical' ? size?.height : size?.width}
+      key={key}
       ref={ref}
+      itemsCount={itemsCount}
       overflowOrientation={overflowOrientation}
     />
   );
@@ -278,15 +283,6 @@ const useOverflow = (
     }
   }, [isStabilized, itemsCount, maxGuess, minGuess, orientation, visibleCount]);
 
-  const [previousItemsCount, setPreviousItemsCount] =
-    React.useState(itemsCount);
-
-  // If itemsCount changes, we have to restart guessing.
-  if (itemsCount !== previousItemsCount) {
-    dispatch({ type: 'itemCountChanged', itemsCount });
-    setPreviousItemsCount(itemsCount);
-  }
-
   // Guess the visible count until stabilized.
   // To prevent flicking, use useLayoutEffect to paint only after stabilized.
   useLayoutEffect(() => {

```
</details> 

---

> This didn't work as the `useLayoutEffect` doesn't call `guessVisibleCount()` when the guess related variables change. So, I didn't implement this.
> 
> The `useLayoutEffect()` may be called a few more times because of it, I think it's alright since once it's stabilized, `guessVisibleCount()` is not called unnecessarily.

I meant to respond to this earlier: Relying on a large deps array of `useLayoutEffect` is usually an anti-pattern. One way of fixing this would be to recursively call `guessVisibleCount` in certain scenarios where we know for sure that we want to re-execute all the logic. This would be worth investigating, maybe in the future, so as to not hold up this PR for too long.